### PR TITLE
utils::is_timeout_exception: Ensure we handle nested exception types

### LINKED ISSUE #9922: commitlog: timed out, during decommission of node
*Installation details*
Kernel version: `5.11.0-1022-aws`
Scylla version (or git commit hash): `4.7.dev-0.20220114.8bcd23fa0 with build-id 991662b202244e747b247a64fa0ddefaab6b53af`
Cluster size: 6 nodes (i3.4xlarge)
Scylla running with shards number (live nodes):
   longevity-100gb-4h-master-db-node-1b686888-1 (54.216.114.122 | 10.0.0.56): 14 shards
   longevity-100gb-4h-master-db-node-1b686888-3 (3.250.14.181 | 10.0.2.117): 14 shards
   longevity-100gb-4h-master-db-node-1b686888-4 (34.250.230.215 | 10.0.0.103): 14 shards
   longevity-100gb-4h-master-db-node-1b686888-5 (63.32.112.153 | 10.0.2.160): 14 shards
   longevity-100gb-4h-master-db-node-1b686888-6 (34.242.73.20 | 10.0.3.107): 14 shards
   longevity-100gb-4h-master-db-node-1b686888-7 (34.247.194.100 | 10.0.2.135): 14 shards
Scylla running with shards number (terminated nodes):
   longevity-100gb-4h-master-db-node-1b686888-2 (3.250.172.53 | 10.0.1.227): 14 shards
OS (RHEL/CentOS/Ubuntu/AWS AMI): `ami-0bf3b97eca8f38134` (aws: eu-west-1)

Test: `longevity-100gb-4h-test`
Test name: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):
- [longevity-100gb-4h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/8584390cb25ea782b18acef4a6d77099be58b0bc/test-cases/longevity/longevity-100gb-4h.yaml)

**Issue description**
During decommission nemesis (taking place on node2), node4 started printing those errors and warnings (820624 of them):

```
2022-01-16T07:43:21+00:00 longevity-100gb-4h-master-db-node-1b686888-4 !     ERR |  [shard 5] storage_proxy - exception during mutation write to 10.0.0.103: std::_Nested_exception<std::runtime_error> (Could not write mutation keyspace1:standard1 (pk{000a3535503450334d383531}) to commitlog): db::commitlog::segment_manager::request_controller_timeout_exception_factory::request_controller_timed_out_error (commitlog: timed out)
...
2022-01-16T07:43:21+00:00 longevity-100gb-4h-master-db-node-1b686888-4 ! WARNING |  [shard 5] storage_proxy - Failed to apply mutation from 10.0.2.135#5: std::_Nested_exception<std::runtime_error> (Could not write mutation keyspace1:standard1 (pk{000a32374d30314f304c5030}) to commitlog): db::commitlog::segment_manager::request_controller_timeout_exception_factory::request_controller_timed_out_error (commitlog: timed out)
```

Restore Monitor Stack command: `$ hydra investigate show-monitor 1b686888-292c-4d17-a4b5-97344b39937f`
Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=1b686888-292c-4d17-a4b5-97344b39937f)
Show all stored logs command: `$ hydra investigate show-logs 1b686888-292c-4d17-a4b5-97344b39937f`

Test id: `1b686888-292c-4d17-a4b5-97344b39937f`

Logs:
db-cluster - [https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/db-cluster-1b686888.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/db-cluster-1b686888.tar.gz)
loader-set - [https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/loader-set-1b686888.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/loader-set-1b686888.tar.gz)
monitor-set - [https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/monitor-set-1b686888.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/monitor-set-1b686888.tar.gz)
97344b39937f/20220116_090328/output-1b686888.log.tar.gz)
sct - [https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/sct-1b686888.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/1b686888-292c-4d17-a4b5-97344b39937f/20220116_090328/sct-1b686888.log.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-100gb-4h-test/438/)

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #9922

storage proxy uses is_timeout_exception to traverse different code paths.
a6202ae079b44da44f402f50001bd2df1cb49673 broke this (because bit rot and
intermixing), by wrapping exception for information purposes.

This adds check of nested types in exception handling, as well as a test
for the routine itself.

### CODE REVIEW

**Review by bhalevy — APPROVED** [review]
LGTM

**Comment by elcallio** [comment]
@bhalevy - added `is_timeout_exception` use in two places missing it.

File: replica/database.cc [hunk 1]
```diff
@@ -1824,11 +1824,10 @@ Future database::update_write_metrics(Future&& f) {
...
         if (f.failed()) {
             ++s->total_writes_failed;
-            try {
-                f.get();
-            } catch (const timed_out_error&) {
+            auto ep = f.get_exception();
+            if (is_timeout_exception(ep)) {
                 ++s->total_writes_timedout;
-                throw;
+                std::rethrow_exception(ep);
```
Comment (bhalevy) [inline replica/database.cc:1830]: Previously non-`timed_out_error` exceptions were not caught and were just propagated back, now they will trigger the assert.

Now we need to rethrow in the `else` case, and then the assert won't be needed.

**Review by bhalevy — CHANGES_REQUESTED** [review]

File: replica/database.cc (hunk 1 above)
Comment (elcallio) [reply replica/database.cc:1830]: Right, sorry, typo.

Comment (elcallio) [reply replica/database.cc]: updated

**Review by bhalevy — APPROVED** [review]

**Review by nyh — APPROVED** [review]

**Pull request closed without merging**